File: Day215_2025/Q215_Leetcode.java
class Q215_Leetcode {
    public int maxTotalFruits(int[][] fruits, int startPos, int k) {
        int n = fruits.length;
        int left = 0, right = 0, sum = 0, ans = 0;
        java.util.function.BiFunction<Integer, Integer, Integer> step = (l, r) -> {
            if (fruits[r][0] <= startPos) {
                return startPos - fruits[l][0];
            } else if (fruits[l][0] >= startPos) {
                return fruits[r][0] - startPos;
            } else {
                return Math.min(Math.abs(startPos - fruits[r][0]), Math.abs(startPos - fruits[l][0])) +
                       fruits[r][0] - fruits[l][0];
            }
        };

        while (right < n) {
            sum += fruits[right][1];
            while (left <= right && step.apply(left, right) > k) {
                sum -= fruits[left][1];
                left++;
            }
            ans = Math.max(ans, sum);
            right++;
        }

        return ans;
    }
}

File: Day215_2025/Q215_Leetcode.py
from typing import List


class Solution:
    def maxTotalFruits(self, fruits: List[List[int]], startPos: int, k: int) -> int:
        n = len(fruits)
        left = 0
        right = 0
        sum_fruits = 0
        ans = 0

        def step(l, r):
            if fruits[r][0] <= startPos:
                return startPos - fruits[l][0]
            elif fruits[l][0] >= startPos:
                return fruits[r][0] - startPos
            else:
                return min(abs(startPos - fruits[r][0]), abs(startPos - fruits[l][0])) + fruits[r][0] - fruits[l][0]

        while right < n:
            sum_fruits += fruits[right][1]
            while left <= right and step(left, right) > k:
                sum_fruits -= fruits[left][1]
                left += 1
            ans = max(ans, sum_fruits)
            right += 1

        return ans
